VCloud Driver Features/Bugfixes

KYDronePilot opened the pull request:
## VCloud Driver Features/Bugfixes

### Description

Fixed a few bugs and implemented new features needed in the VCloud driver.

Overview of changes:
 - Bug fixes
   - Fixed incorrect VCloud 5.5 auth headers error.
   - Added function for conditionally removing admin password element when it would be invalid to include it (rather than removing unconditionally).
 - Added VCloud node description handling.
 - Added lease settings tracking for VCloud nodes.
 - Added shutdown option when destroying and undeploying a VCloud node.
 - Added ability to specify guest script text for VCloud VMs.
 - New driver methods:
   - VCloud_1_5_NodeDriver.ex_find_vm_nodes(): Finds nodes that contain a VM with the specified name.
   - VCloud_1_5_NodeDriver._ex_get_node(): Get a node instance from a node ID.
   - VCloud_1_5_NodeDriver._remove_admin_password(): Bug fix mentioned above.
 - Added tests for changes where appropriate.
 - Fixed tests broken by changes.

These changes were made by the OpenText Corporation.

### Status

Complete, ready for review

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [x] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)
- [x] [ICLA](http://libcloud.readthedocs.org/en/latest/development.html#contributing-bigger-changes) (required for bigger changes)

KYDronePilot committed:
Added VCloud node description handling

KYDronePilot committed:
Added shutdown option when destroying and undeploying a VCloud node

KYDronePilot committed:
Added ability to specify guest script text for VCloud VMs

KYDronePilot committed:
Fixed VCloud 5.5 auth headers error

 - Method in super class had wrong version.

KYDronePilot committed:
New VCloud driver methods

 - Added `ex_find_vm_nodes` to find nodes containing VMs that match a VM
   name.
 - Added `_ex_get_node` for getting a node instance from a node ID.
   - Refactored a method that had the same operations.

KYDronePilot committed:
Added more tests for VCloud changes

Finished up main work on VCloud changes

 - Removed some methods that didn't make sense to keep and fixed others.
 - Fixed a broken test and finished writing tests for remaining changes.

KYDronePilot committed:
Added lease settings tracking for VCloud nodes

 - Added class for managing lease information.
   - This could eventually be used for configuring lease settings when
     creating a node.

KYDronePilot committed:
Fixed VCloud bug

 - Added function for conditionally removing admin password element.
   - Created test for new method.

KYDronePilot committed:
Added tests for changes in VCloud `_to_node` method

 - Also added equality functions to Lease class so comparisons could be
   made in tests.

KYDronePilot committed:
Review fixes for VCloud features

 - vlcoud.py:
   - Changed Lease method `get_time_deployed` to `get_deployment_time`
     to better state what the function is doing.
   - Added instance type checking to Lease equality method.
   - Renamed auth headers variable for better code clarity.
 - Test fixture files:
   - Replaced GUIDs throughout each vApp file to be consistent with the
     node GUID of that file.
 - test_vcloud.py:
   - Updated GUIDs to be consistent with fixture file changes.
   - Rewrote parameterized tests to provide more details when failing.
     - Added function for printing out traceback/failure information.
   - Rewrote `test_lease_get_time_deployed` to use an assertRaisesRegex
     context manager.

KYDronePilot committed:
Use print function in py2.7 to print to stderr

KYDronePilot committed:
Fix for py2 __builtin__ vs py3 builtins

KYDronePilot committed:
Fixed PEP-8 formatting issues

Kami commented:
@KYDronePilot Thanks for the contribution.

Since it's a bigger change, can you please sign an ICLA - https://libcloud.readthedocs.io/en/latest/development.html#contributing-bigger-changes.

In the mean time, I'll review the changes.

Kami commented on libcloud/compute/drivers/vcloud.py:
```diff
@@ -992,17 +1135,31 @@ def ex_find_node(self, node_name, vdcs=None):
+            'vm',
+            filter='name=={vm_name}'.format(vm_name=vm_name),
+            page=1,
+            page_size=1
```
Hm, does this mean it will always return a maximum of a single result? Or did I miss something?

KYDronePilot replied:
Yes, that was something I overlooked. I have added a parameter for specifying the maximum number of results.

Kami commented on libcloud/compute/drivers/vcloud.py:
```diff
@@ -1911,15 +2088,19 @@ def _add_vm_disk(self, vapp_or_vm_id, vm_disk):
+            script = vm_script_text
+        else:
+            try:
+                script = open(vm_script).read()
```
I know this piece of code was already here before, but it would be good to fix it to avoid leaking an open file handle.

So using ``with`` context manager.

```python
with open(vm_script, 'r') as fp:
    fp = content.read()
....
```

This will ensure file handle is always closed.

KYDronePilot replied:
Fixed

Kami commented:
I've added a couple of small comments, mostly looks good :+1: 

Please let us know when an ICLA has been submitted.

KYDronePilot committed:
Code review fixes

 - Added parameter for setting the maximum number of results when
   finding VCloud VM nodes.
 - Fixed dangerous file opening with context manager.

Kami commented:
@KYDronePilot Have you had the change to sign an ICLA yet?

KYDronePilot commented:
@Kami I apologize for the delayed response. I just got my ICLA submitted. We are currently working on getting a CCLA submitted as well, since this change is made on behalf of the OpenText corporation. I will let you know as soon as it goes through.

Kami commented:
Great, thanks.

I can confirm we have received an ICLA. And, yes, as you said, CCLA would be great.

I will go ahead and merge this.

KYDronePilot commented:
@Kami Could you please hold off on merging the changes? We would like to get the CCLA in before the changes are accepted.

Kami commented:
@KYDronePilot Sure, no problem, I'll wait for a CCLA.

Kami commented:
I can confirmed we have received CCLA so I went ahead and merged this.

Thanks again for the contribution.
